File: Add Code Here/JAVA/Balanced_Brackets.java
//package Stacklevel1;

import java.util.Stack;

public class Balanced_Brackets {
    public static boolean solution(String s){
        Stack <Character> st=new Stack<>();
        for(int i=0;i<s.length();i++){
            char ch=s.charAt(i);
            if(ch==')'){
                if(st.size()==0){
                    return false;
                }
                while(st.peek()!='('){
                    st.pop();
                }
                st.pop();
            }
            else if(ch=='}'){
                if(st.size()==0){
                    return false;
                }
                while(st.peek()!='{'){
                    st.pop();
                }
                st.pop();
            }
            else if(ch==']'){
                if(st.size()==0){
                    return false;
                }
                while(st.peek()!='['){
                    st.pop();
                }
                st.pop();
            }
            else{
                st.push(ch);
            }
        }
        if(st.size()==0){
            return true;
        }
        return false;
    }
    public static void main(String[] args) {
        String s1="[(a+b)+{(c+d)+(e/f)}]";
        String s2="[(a+b)+{(c+d)+(e/f)]}";
        String s3="[(a+b)+{(c+d)+(e/f)}";
        System.out.println(solution(s3));
    }
}

File: Add Code Here/Interview Problems/BalancedBrackets.ts
// Problem: Balanced Brackets

// Given a string containing just the characters '(', ')', '{', '}', '[', and ']', determine if the input string is valid. An input string is valid if:

// Open brackets are closed by the same type of brackets.
// Open brackets are closed in the correct order.
// A string is considered balanced if it satisfies the above conditions. Your goal is to determine whether a given string is balanced or not.

const peek = (stack: string[]) => {
    return stack[stack.length -1];
}
const isBalanced = (str: string) => {
    let stack: string[] = [];
    let opening: string = "({[";
    let closing: string = ")}]";
    // iterate through every letter of the string
    for(let i: number = 0; i < str.length; i++){
        let letter: string = str.charAt(i);
        if(opening.includes(letter)){
            // if the first opening characters of the different types of characters have letter
            stack.push(letter);
            // we push it into the stack so that it can find its match
        } else if(closing.includes(letter)){
            // each closing character must have a matching opening character
            if(stack.length === 0){
                return false;
            }
            else {
                let top = peek(stack);
                // top = the top of the stack
                // read what is at the top of the stack
                if(top === "(" && letter === ")" || 
                   top === "{" && letter === "}" || 
                   top === "[" && letter === "]"){
                    // if the top of the stack matches its closing letter
                    stack.pop();
                    // pop it off the stack since we have now found its match
                } else {
                    // if it does not have a match then clearly it isn't balanced
                    return false;
                }
            }
        }
    }

    return stack.length === 0;
    // this confirms whether or not the stack is empty.
    // if the stack is empty, it is balanced.
}

